scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -619,163 +741,336 @@ class ProcessMonitor(ScalyrMonitor):
-            self.__gathers.append(StatReader(self.__pid, self.__id, self._logger))
-            self.__gathers.append(StatusReader(self.__pid, self.__id, self._logger))
-            self.__gathers.append(IoReader(self.__pid, self.__id, self._logger))
-            self.__gathers.append(FileDescriptorReader(self.__pid, self.__id, self._logger))
-        # TODO: Re-enable these if we can find a way to get them to truly report
-        # per-app statistics.
-        #        self.gathers.append(NetStatReader(self.pid, self.id, self._logger))
-        #        self.gathers.append(SockStatReader(self.pid, self.id, self._logger))
+    def _reset_absolute_metrics(self):
+        """
+        At the beginning of each process metric calculation, the absolute (non-cumulative) metrics
+        need to be overwritten to the combined process(es) result. Only the cumulative
+        metrics need the previous value to calculate delta. We should set the
+        absolute metric to 0 in the beginning of this "epoch"
+        """
 
-    def gather_sample(self):
-        """Collect the per-process metrics for the monitored process.
+        for pid, process_metrics in self.__metrics_history.items():
+            for _metric, _metric_values in process_metrics.items():
+                if not _metric.is_cumulative:
+                    self.__aggregated_metrics[_metric] = 0
 
-        If the process is no longer running, then attempts to match a new one.
+    def _calculate_aggregated_metrics(self, running_pids):
+        """
+        Calculates the aggregated metric values based on the current running processes
+        and the historical metric record
+        @param running_pids: list of running process ids
+        @type running_pids: list
         """
-        if self.__pid is not None and not self.__is_running():
-            self.__set_process(None)
 
-        if self.__pid is None:
-            self.__set_process(self.__select_process())
+        # using the historical values, calculate the aggregate
+        # there are two kinds of metrics:
+        # a) cumulative metrics - only the delta of the last 2 recorded values is used (eg cpu cycles)
+        # b) absolute metrics - the last absolute value is used
+
+        running_pids_set = set(running_pids)
+
+        for pid, process_metrics in self.__metrics_history.items():
+            for _metric, _metric_values in process_metrics.items():
+                if not self.__aggregated_metrics.get(_metric):
+                    self.__aggregated_metrics[_metric] = 0
+                if _metric.is_cumulative:
+                    if pid in running_pids_set:
+                        if len(_metric_values) > 1:
+                            # only report the cumulative metrics for more than one sample
+                            self.__aggregated_metrics[_metric] += (_metric_values[-1] - _metric_values[-2])
+                        else:
+                            self.__aggregated_metrics[_metric] = 0
```

Comment:
I don't think you should be setting this to zero... that means anytime a new process comes along, we would be resetting everything to zero.  Instead, just don't do anything if `len(_metric_values) <= 1`